jetstream_test.go

```diff
@@ -0,0 +1,86 @@
+package nats
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/nats-io/nats-server/v2/server"
+	natsserver "github.com/nats-io/nats-server/v2/test"
+)
+
+func startJetStream(t *testing.T) *server.Server {
+	sopts := natsserver.DefaultTestOptions
+	sopts.JetStream = true
+	sopts.Port = -1
+
+	td, err := ioutil.TempDir("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	srv := natsserver.RunServer(&server.Options{JetStream: true, StoreDir: td})
+	_, err = srv.GlobalAccount().AddStream(&server.StreamConfig{
+		Name:     "TEST",
+		Subjects: []string{"js.in.test"},
+		Storage:  server.MemoryStorage,
+	})
+	if err != nil {
+		t.Fatalf("stream create failed: %v", err)
+	}
+
+	return srv
+}
+
+func TestJetStreamPublish(t *testing.T) {
+	srv := startJetStream(t)
+	defer os.RemoveAll(srv.JetStreamConfig().StoreDir)
+	defer srv.Shutdown()
+
+	nc, err := Connect(srv.ClientURL())
+	if err != nil {
+		t.Fatalf("connect failed: %v", err)
+	}
+
+	err = nc.Publish("js.in.test", []byte("hello"), ExpectAckFromStream("TEST"))
+	if err != nil {
+		t.Fatalf("publish failed: %v", err)
+	}
+
+	err = nc.Publish("js.in.test", []byte("hello"), ExpectAckFromStream("OTHER"))
+	if err == nil {
+		t.Fatalf("expected an error but got none")
+	}
+}
+
+func TestMsg_ParseJSMsgMetadata(t *testing.T) {
+	msg := &Msg{Reply: "$JS.ACK.ORDERS.NEW.1.2.3.1587466354254920000"}
+	err := msg.ParseJSMsgMetadata()
```

Comment:
@derekcollison following up from our call, we thought parsing these into headers on demand in the client would be nice - it's ok, I guess but things like `JS-Delivered` being a string is very annoying.

Do you think we should do this anyway and then add some accessors on `msg` to fetch and convert as a helper?

Reply:
ok let's add helpers then instead.

Reply:
OK, done that, seems better.

Next to see what we can do for consumers